### ISSUE TITLE
Generics and issubclass TypeGuard

### ISSUE THREAD

**Issue by tamuhey** [microsoft/pyright#2096]
```python
from typing import Type, TypeVar

T = TypeVar("T")


def f(x: Type[T]) -> T:
    if issubclass(x, bool):
        return True  # ERROR: error: Expression of type "Literal[True]" cannot be assigned to return type "T@f"
    raise ValueError()
```

1. I think `pyright` accepts this code, but I don't know it is technically possible. 
2. Is there any other guards in which I can narrow down `T` to `bool`?

**Comment by erictraut** [comment]
This is correct behavior. When using a regular (not constrained) type variable, the type must match exactly. In your sample, the code is validating that `x` is a subclass of `bool`, but it could be some derived class, in which case it wouldn't be appropriate to return a `bool`. You can return `x()` without generating a type error.

I'm not sure what you're trying to do in this code since it's a stripped-down example. Is your intent to return True if the specified type is a `bool`? If so, then why is it necessary for the function to be generic? You could simply do the following:
```python
def is_bool(x: type) -> bool:
    if issubclass(x, bool):
        return True
    raise ValueError()
```

Or if you want to use PEP 647, you could do:
```python
def is_bool2(x: type) -> TypeGuard[Type[bool]]:
    return issubclass(x, bool)
```

But that's kind of redundant because pyright already handles `issubclass` as a type guard.

**Comment by tamuhey** [comment]
@erictraut Thank you for your response. Basically, I want to do the following:

```python
from typing import List, Type, TypeVar
T = TypeVar("T", str, int, List[str], bool)

def parse_commandline_arg(ty: Type[T], values: List[str]) -> T:
    if ty == bool:
        return len(values) > 0
    elif ty == str:
        return values[0]
    elif ty == list:
        return values
    elif ty == int:
        return int(values[0])
    raise ValueError()
```

**Comment by JelleZijlstra** [comment]
Note though that bool is marked as final (https://github.com/python/typeshed/blob/master/stdlib/builtins.pyi#L684) so there can be no subclasses of it. That means the code in the original example is actually safe.

**Comment by erictraut** [comment]
The top example is arguably safe in this very specific case, but a type checker (including pyright and mypy) will still mark it as an error because the return expression `True` has no relationship with the type being narrowed in the type guard.

The other sample immediately above is clearly in error. The use of a type variable here makes no sense because the outputs have no relation to the input type `T`.

The function should be:
```python
def parse_commandline_arg(ty: type, values: List[str]) -> Union[str, int, List[str], bool]:
    ...
```
